lib/linked_list.rb

```diff
@@ -14,75 +14,171 @@ def initialize(value, next_node = nil)
 class LinkedList
     def initialize
       @head = nil # keep the head private. Not accessible outside this class
+      @tail = nil
     end
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(value)
-      raise NotImplementedError
+      if @head.nil?
+        @head = @tail = Node.new(value)
+      else
+        new_node = Node.new(value, @head)
+        @head = new_node
+      end
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def search(value)
-      raise NotImplementedError
+      return false if @head.nil?
+      current = @head
+
+      until current.nil?
+        if current.data == value
+          return true
+        end
+        current = current.next
+      end
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_max
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      max_value = @head.data
+
+      until current.nil?
+        if current.data > max_value
+          max_value = current.data
+        end
+        current = current.next
+      end
+
+      return max_value
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      min_value = @head.data
+
+      until current.nil?
+        if current.data < min_value
+          min_value = current.data
+        end
+        current = current.next
+      end
+
+      return min_value
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def length
-      raise NotImplementedError
+      return 0 if @head.nil?
+      current = @head
+      count = 0
+
+      until current.nil?
+        count += 1
+        current = current.next
+      end
+
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def get_at_index(index)

```

Comment:
👍 , The space complexity is O(1) because you're not making a new list with this method.